### PR TITLE
Introduce decompressor interface for ORC reader

### PR DESCRIPTION
**Pull request by martint**

### CODE REVIEW

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcDecompressor.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * limitations under the License.
+ */
+package com.facebook.presto.orc;
+
+public interface OrcDecompressor
+{
+    int decompress(byte[] input, int offset, int length, Buffer output)
+            throws OrcCorruptionException;
+
+    interface Buffer
```
Comment (dain) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcDecompressor.java]: `OutputBuffer`

File: presto-orc/src/main/java/com/facebook/presto/orc/checkpoint/DecimalStreamCheckpoint.java [hunk 2]
```diff
@@ -25,9 +24,9 @@
 {
     private final long inputStreamCheckpoint;
 
-    public DecimalStreamCheckpoint(CompressionKind compressionKind, ColumnPositionsList positionsList)
+    public DecimalStreamCheckpoint(boolean decompressor, ColumnPositionsList positionsList)
```
Comment (dain) [inline presto-orc/src/main/java/com/facebook/presto/orc/checkpoint/DecimalStreamCheckpoint.java]: `compressed`?

File: presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java [hunk 3]
```diff
@@ -254,69 +253,7 @@ public String toString()
                 .add("source", source)
                 .add("compressedOffset", compressedSliceInput.position())
                 .add("uncompressedOffset", current == null ? null : current.position())
-                .add("compression", compressionKind)
+                .add("decompressor", decompressor)
```
Comment (dain) [inline presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java]: maybe add a `toString` to the compressor implementations which returns `snappy` or `zlib` and then change this to `decompressor.map(Object::toString).orElse("none")`

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcSnappyDecompressor.java [hunk 4]
```diff
@@ -31,10 +32,10 @@ public OrcSnappyDecompressor(int maxBufferSize)
     public int decompress(byte[] input, int offset, int length, Buffer output)
             throws OrcCorruptionException
     {
-        int uncompressedLength = Snappy.getUncompressedLength(input, offset);
+        int uncompressedLength = SnappyDecompressor.getUncompressedLength(input, offset);
         checkArgument(uncompressedLength <= maxBufferSize, "Snappy requires buffer (%s) larger than max size (%s)", uncompressedLength, maxBufferSize);
-        byte[] buffer = output.initialize(uncompressedLength);
 
-        return Snappy.uncompress(input, offset, length, buffer, 0);
+        byte[] buffer = output.initialize(uncompressedLength);
```
Comment (dain) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcSnappyDecompressor.java]: Add `SIZE_OF_LONG` to the buffer, which helps speed up the snappy decompressor.

File: presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java [hunk 5]
```diff
@@ -254,69 +253,7 @@ public String toString()
                 .add("source", source)
                 .add("compressedOffset", compressedSliceInput.position())
                 .add("uncompressedOffset", current == null ? null : current.position())
-                .add("compression", compressionKind)
+                .add("decompressor", decompressor.map(Object::toString).orElse("none"))
```
Comment (findepi) [inline presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java]: Minor: `decompressor.orElse("none")` would be equivalent.

Comment (dain) [inline presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java]: The argument in orElse must be the same type

**Pull request merged**